subt/test_subt.py

```diff
@@ -1,7 +1,173 @@
+import collections
+import datetime
+import inspect
+import math
 import unittest
+
 from unittest.mock import MagicMock
+from threading import Thread, RLock
 
+from osgar.lib import quaternion
+from osgar.bus import Bus
 from subt.main import SubTChallenge
+from subt.trace import distance3D
+
+try:
+    from colored import stylize, fg
+except ImportError:
+    def stylize(what, style):
+        return what
+    def fg(color_name):
+        return ""
+
+import sys
+if '-v' not in sys.argv or '--verbose' not in sys.argv:
+    def print(*args):
+        pass
+
+def log(*args):
+    up = inspect.currentframe().f_back
+    caller = up.f_code.co_name
+    caller_class = up.f_locals['self'].__class__
+    color_name = getattr(caller_class, caller).color
+    print(stylize(f"{caller_class.__name__}.{caller}", fg(color_name)), *args)
+
+
+def clip(min_value, value, max_value):
+    return max(min(value, max_value), min_value)
+
+
+class Color:
+    def __getattr__(self, color_name):
+        def a(method):
+            method.color = color_name
+            return method
+        return a
+
+
+color = Color()
+
+
+class SimLogger:
+
+    def __init__(self):
+        self.lock = RLock()
+        self.start_time = datetime.datetime.now(datetime.timezone.utc)
+        self.names = ["sys"]
+
+    def register(self, name):
+        self.names.append(name)
+        return len(self.names)-1
+
+    @color.dodger_blue_1
+    def write(self, stream_id, data, dt=None):
+        #log(self.names[stream_id], data)
+        dt = datetime.datetime.now(datetime.timezone.utc) - self.start_time
+        return dt
+
+
+class Simulation:
```

Comment:
this part seems too big  ... I would probably keep it separate from unittest??

Reply:
yeah, that's right - any idea about the name? `subt/simulation.py`?

Reply:
that sounds reasonable ... i.e. yes